File: ts-core/src/net/buffer.ts
import {
    DeserializeError,
    DeserializeResult,
    Deserializer,
    Reader,
    SerializeError,
    SerializeResult,
    Serializer,
    Writer,
} from "@core/serde";
import { Err, Ok, Result } from "oxide.ts";

export class BufferOutOfBoundsException extends Error {
    constructor(buffer: Uint8Array, requested: number) {
        super(`Buffer out of bounds: ${buffer.length} < ${requested}`);
    }
}

export class BufferWriter implements Writer {
    #buffer: Uint8Array;
    #offset = 0;

    constructor(buffer: Uint8Array) {
        this.#buffer = buffer;
    }

    static serialize<S extends Serializer>(serializer: S): Result<Uint8Array, SerializeError> {
        const length = serializer.serializedLength();
        const writer = new BufferWriter(new Uint8Array(length));
        return serializer.serialize(writer).map(() => writer.unwrapBuffer());
    }

    #requireBounds(requested: number): SerializeResult {
        const expected = this.#offset + requested;
        const actual = this.#buffer.length;
        return expected <= actual ? Ok(undefined) : Err(new Error(`Buffer out of bounds: ${expected} < ${actual}`));
    }

    writeByte(byte: number): SerializeResult {
        return this.#requireBounds(1).map(() => {
            this.#buffer[this.#offset++] = byte;
        });
    }

    writeBytes(bytes: Uint8Array): SerializeResult {
        return this.#requireBounds(bytes.length).map(() => {
            this.#buffer.set(bytes, this.#offset);
            this.#offset += bytes.length;
        });
    }

    unwrapBuffer(): Uint8Array {
        if (this.#offset !== this.#buffer.length) {
            throw new Error(`Buffer not fully written: ${this.#offset} < ${this.#buffer.length}`);
        }
        return this.#buffer;
    }
}

export class BufferReader implements Reader {
    #buffer: Uint8Array;
    #offset = 0;

    constructor(buffer: Uint8Array) {
        this.#buffer = buffer;
    }

    static deserialize<T>(deserializer: Deserializer<T>, buffer: Uint8Array): DeserializeResult<T> {
        return deserializer.deserialize(new BufferReader(buffer));
    }

    #requireBounds(requested: number): DeserializeResult<void> {
        const expected = this.#offset + requested;
        const actual = this.#buffer.length;
        return expected <= actual ? Ok(undefined) : this.notEnoughBytesError(expected, actual);
    }

    readByte(): DeserializeResult<number> {
        return this.#requireBounds(1).map(() => this.#buffer[this.#offset++]);
    }

    readBytes(length: number): DeserializeResult<Uint8Array> {
        return this.#requireBounds(length).map(() => {
            const bytes = this.#buffer.subarray(this.#offset, this.#offset + length);
            this.#offset += length;
            return bytes;
        });
    }

    notEnoughBytesError(expected: number, actual: number): Err<DeserializeError> {
        return Err(new DeserializeError.NotEnoughBytesError(this.#buffer, expected, actual));
    }

    invalidValueError(name: string, value: unknown): Err<DeserializeError> {
        return Err(new DeserializeError.InvalidValueError(this.#buffer, name, value));
    }

    readCount(length: number): DeserializeResult<Uint8Array> {
        return this.#requireBounds(length).map(() => {
            const bytes = this.#buffer.subarray(this.#offset, this.#offset + length);
            this.#offset += length;
            return bytes;
        });
    }

    readRemainingBytes(): DeserializeResult<Uint8Array> {
        return Ok(this.readRemaining());
    }

    readBytesMax(maxLength: number): Uint8Array {
        const readable = this.#buffer.length - this.#offset;
        const length = Math.min(maxLength, readable);
        return this.readCount(length).unwrap();
    }

    readRemaining(): Uint8Array {
        const bytes = this.#buffer.subarray(this.#offset);
        this.#offset = this.#buffer.length;
        return bytes;
    }

    remainingLength(): number {
        return this.#buffer.length - this.#offset;
    }

    initialized(): BufferReader {
        return new BufferReader(this.#buffer);
    }

    subReader(): BufferReader {
        return new BufferReader(this.#buffer.subarray(this.#offset));
    }
}

File: py-core/py_core/serde/buffer.py
from typing import Never, Protocol


class NotEnoughSpaceException(Exception):
    pass


class UnwrapUnfinishedBufferException(Exception):
    pass


class Writer(Protocol):
    def write_byte(self, value: int) -> None: ...

    def write_bytes(self, data: bytes) -> None: ...


class BufferWriter(Writer):
    _buffer: bytearray
    _index: int

    def __init__(self, length: int = 0):
        self._buffer = bytearray(length)
        self._index = 0

    def _assert_enough_space(self, n: int) -> None:
        if self._index + n > len(self._buffer):
            raise NotEnoughSpaceException()

    def write_byte(self, value: int) -> None:
        self._assert_enough_space(1)
        self._buffer[self._index] = value
        self._index += 1

    def write_bytes(self, data: bytes) -> None:
        self._assert_enough_space(len(data))
        self._buffer[self._index : self._index + len(data)] = data
        self._index += len(data)

    def unwrap_buffer(self) -> bytearray:
        if self._index != len(self._buffer):
            raise UnwrapUnfinishedBufferException()
        return self._buffer


class NotEnoughBytesException(Exception):
    buffer: bytes
    expected: int
    actual: int

    def __init__(self, buffer: bytes, expected: int, actual: int):
        self.buffer = buffer
        self.expected = expected
        self.actual = actual
        super().__init__(f"Expected {expected} bytes, but got {actual}")


class InvalidValueException(Exception):
    buffer: bytes
    name: str
    value: object

    def __init__(self, buffer: bytes, name: str, value: object):
        self.buffer = buffer
        self.name = name
        self.value = value
        super().__init__(f"Invalid value for {name}: {value}")


class Reader(Protocol):
    def read_byte(self) -> int: ...

    def read_bytes(self, n: int) -> bytes: ...

    def read_remaining(self) -> bytes: ...

    def raise_invalid_value_error(self, name: str, value: object) -> Never: ...


class BufferReader(Reader):
    _buffer: bytes
    _index: int

    def __init__(self, buffer: bytes):
        self._buffer = buffer
        self._index = 0

    def _assert_enough_bytes(self, n: int) -> None:
        if self._index + n > len(self._buffer):
            raise NotEnoughBytesException(
                self._buffer, n, len(self._buffer) - self._index
            )

    def read_byte(self) -> int:
        self._assert_enough_bytes(1)
        value = self._buffer[self._index]
        self._index += 1
        return value

    def read_bytes(self, n: int) -> bytes:
        self._assert_enough_bytes(n)
        value = self._buffer[self._index : self._index + n]
        self._index += n
        return value

    def read_remaining(self) -> bytes:
        value = self._buffer[self._index :]
        self._index = len(self._buffer)
        return value

    def raise_invalid_value_error(self, name: str, value: object) -> Never:
        raise InvalidValueException(self._buffer, name, value)

    def remaining_length(self) -> int:
        return len(self._buffer) - self._index

    def sub_reader(self) -> "BufferReader":
        return BufferReader(self._buffer[self._index :])
